src/eve/utils.py

```diff
@@ -281,3 +401,680 @@ def reset_sequence(cls, start: int = 1) -> None:
+
+    def __init__(self, it: Union[Iterable[T], Iterator[T]]) -> None:
+        if not isinstance(it, collections.abc.Iterator):
+            raise ValueError(f"Invalid iterator instance: '{it}'.")
+        super().__setattr__("iterator", it.iterator if isinstance(it, XIterator) else it)
+
+    def __getattr__(self, name: str) -> Any:
+        # Forward special methods to wrapped iterator
+        if name.startswith("__") and name.endswith("__"):
+            return getattr(self.iterator, name)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        raise TypeError(f"{type(self).__name__} is immutable.")
+
+    def __next__(self) -> T:
+        return next(self.iterator)
+
+    def map(self, func: AnyCallable) -> XIterator[Any]:  # noqa  # A003: shadowing a python builtin
+        """Apply a callable to every iterator element (equivalent to ``map(func, self)``).
+
+        For detailed information check :func:`map` reference.
+
+        Examples:
+            >>> it = xiter(range(3))
+            >>> list(it.map(str))
+            ['0', '1', '2']
+
+            >>> it = xiter(range(3))
+            >>> list(it.map(lambda x: -x).map(str))
+            ['0', '-1', '-2']
+
+            If the callable requires additional arguments, ``lambda`` of :func:`functools.partial`
+            functions can be used:
+
+            >>> def times(a, b):
+            ...     return a * b
+            >>> times_2 = functools.partial(times, 2)
+            >>> it = xiter(range(4))
+            >>> list(it.map(lambda x: x + 1).map(times_2))
+            [2, 4, 6, 8]
+
+            Curried functions generated by :func:`toolz.curry` will also work as expected:
+
+            >>> @toolz.curry
+            ... def mul(x, y):
+            ...     return x * y
+            >>> it = xiter(range(4))
+            >>> list(it.map(lambda x: x + 1).map(mul(5)))
+            [5, 10, 15, 20]
+
+        """
+        if not callable(func):
+            raise ValueError(f"Invalid function or callable: '{func}'.")
+        return XIterator(map(func, self.iterator))
+
+    def filter(  # noqa  # A003: shadowing a python builtin
+        self, func: Callable[..., bool]
+    ) -> XIterator[T]:
+        """Filter elements with callables (equivalent to ``filter(func, self)``).
+
+        For detailed information check :func:`filter` reference.
+
+        Examples:
+            >>> it = xiter(range(4))
+            >>> list(it.filter(lambda x: x % 2 == 0))
+            [0, 2]
+
+            >>> it = xiter(range(4))
+            >>> list(it.filter(lambda x: x % 2 == 0).filter(lambda x: x > 1))
+            [2]
+
+
+        Notes:
+            `lambdas`, `partial` and `curried` functions are supported (see :meth:`map`).
+
+        """
+        if not callable(func):
+            raise ValueError(f"Invalid function or callable: '{func}'.")
+        return XIterator(filter(func, self.iterator))
+
+    def filter_by_type(self, *types: Type) -> XIterator[T]:
```

Comment:
as you proposed
```suggestion
    def by_type(self, *types: Type) -> XIterator[T]:
```

Reply:
or `filter_type`?

Reply:
In the end I used `if_isinstance` which is also short and closer to the actual implementation